### ISSUE TITLE
Runtime / user-granted roles and scopes

### ISSUE THREAD

**Issue by minrk** [jupyterhub/jupyterhub#3858]
### Proposed change

For sharing and user-initiated collaboration, I think we want the ability for users to _grant_ other users permissions. Right now, to collaborate, admins must grant minimum sharing permissions _at all times_. This works for protyping purposes, and single-group deployments, but not in general

For simplicity, the first implementation of RBAC only allows roles to be defined in config. That means we always reconcile all role assignments to reflect config - roles cannot be created without restarting the hub, and their definitions only reside in config.

Example use cases:

1. start a server (e.g. a dashboard) and make it accessible by all users, unlike the user's default server (https://discourse.jupyter.org/t/refining-server-scopes-in-role-definitions/13760) 
2. user grants limited (e.g. read-only) collaboration access to their current server, _without_ requiring that the other user have that access all the time. The user should be able to revoke permissions they have granted.

The challenges are in reconciliation with config:

1. if we can grant access during runtime, we can't revert to config at each restart. We have to recognize what changes are from changed config and what changes are from runtime, priorities, etc.
2. what's the difference at startup between a role deleted from config (should be deleted) and a role created during runtime (should be left alone)?

### Who would use this feature?

Deployments that want selective and user-controlled collaboration.

### (Optional): Suggest a solution

- REST API for managing roles, role assignments
- need a way to track role changes from config vs runtime to avoid making a mess during hub startup. A single flag on the role might be the simplest way to track - config-managed roles could be read-only via API.
- need to track roles 'owned' by users
- Users may _create_ roles, granting other users permissions
- something like `oauth_allowed_roles`, but for users to govern which permissions they can grant others

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/refining-server-scopes-in-role-definitions/13760/4

**Comment by ktaletsk** [comment]
> REST API for managing roles, role assignments

This would help a lot! Currently I have to run wrapper to restart JupyterHub to add individual sharing roles for each user. https://github.com/ktaletsk/jupyterhub-rtc-config-wrapper

**Comment by minrk** [comment]
#4594 is now complete and ready for testing. I still think runtime-defined roles makes sense, so it won't close this issue, but it at least addresses the _primary_ use case.

**Comment by krassowski** [comment]
Was having a `manage_roles` which would mirror `manage_groups` but for roles discussed before?

https://github.com/jupyterhub/jupyterhub/blob/5033effffb89793bfa0af62d9153a9ad504bf411/jupyterhub/handlers/base.py#L843-L856

Specifically, I am asking about a use case where roles are defined in the `Authenticator`'s database which makes using `c.JupyterHub.load_roles` not very hand because the `Authenticator` may not be ready on the time of configuration. This is in contrast to the existing proposal of exposing the REST API for role management (https://github.com/jupyterhub/jupyterhub/pull/3980).

**Comment by minrk** [comment]
Yes, this is proposed in #4050 and I think makes sense. That PR is unfinished because it still needs to address the Hub startup issue reconciling role membership, but if someone can pick that up it should be doable. I believe it should work to skip initializing roles altogether when `Authenticator.manage_roles` is True (it should be mutually exclusive with `load_roles`).

**Comment by krassowski** [comment]
> I believe it should work to skip initializing roles altogether when `Authenticator.manage_roles` is True (it should be mutually exclusive with `load_roles`).

Should it skip `load_roles` but still load the default roles (user, admin, server, token)?

**Comment by minrk** [comment]
> Should it skip load_roles but still load the default roles (user, admin, server, token)?

Yes, it should define the default roles, but load no role assignments. Or at least no _user_ role assignments.

Some questions to answer:

- what about service roles?
- how/when are roles defined and their scopes loaded? Should the Authenticator have its own role-initialization method called during startup?
